### PR TITLE
Tpetra: Fix #5279 (add test for checking a pointer to see if it's CUDA-accessible)

### LINKED ISSUE #5279: Tpetra: Check input Views to make sure pointer matches memory space
## Enhancement
@trilinos/tpetra 

Context: Run-time errors in Panzer CUDA build.  See discussion [starting here](https://github.com/trilinos/Trilinos/issues/5179#issuecomment-496962330).

Whenever a Tpetra object (Map, MultiVector, CrsGraph, CrsMatrix) takes a Kokkos::View or Kokkos::DualView from the user, use the CUDA Runtime API to check the pointers.  Make sure that if the user claims that the View is a device View, then the pointer is a valid device pointer.

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/tpetra @bathmatt @vbrunini 

## Description

Add a Tpetra implementation detail function to test whether a given nonnull pointer is CUDA-accessible.  This is a partial fix for #5279.  A full fix would involve finding all the Tpetra object constructors that take `Kokkos::View` or `Kokkos::DualView` from the user, and using this function in those constructors.  I'll open a separate issue for that.

## Motivation and Context

Help debug run-time errors in Panzer: see [discussion here](https://github.com/trilinos/Trilinos/issues/5179#issuecomment-496962330).

@vbrunini suggested that `Kokkos::View`'s constructor (the overloads that take a raw pointer from the user) could use this function, at least in debug mode.  @crtrott , what do you think?

## Related Issues

* Closes #5279 
* Related to #5179 

## How Has This Been Tested?

CUDA 9.2 build with GCC 7.2.0 (x86) and OpenMPI 1.10.1.

### CODE REVIEW

**Comment by bathmatt** [comment]
@mhoemmen do you need some flag to turn this on..??

**Comment by mhoemmen** [comment]
@bathmatt wrote:
> @mhoemmen do you need some flag to turn this on..??

This is only the first step.  Once I get this PR through, I will make Tpetra classes like Map, MultiVector, CrsGraph, and CrsMatrix use this function.  Their constructors will use the `TPETRA_DEBUG` environment variable to decide whether to check.  If it's not expensive, I may just check by default.

**Comment by mhoemmen** [comment]
@trilinos/framework Something appears to be wrong with tests.

**Comment by bathmatt** [comment]
OK, so there is no reason to test this yet?  It won't show anything?

**Comment by mhoemmen** [comment]
@bathmatt wrote:
> OK, so there is no reason to test this yet? It won't show anything?

That's right, this first part is just the function for testing itself, not the use of that function in Tpetra objects.

**Comment by jwillenbring** [comment]
> Something appears to be wrong with tests.

I am looking at it now.

**Comment by mhoemmen** [comment]
CUDA failure:
```

0. Utils_CheckMemoryType_UnitTest ... 
 Starting CheckMemoryType test
  Test raw integer on host
   memorySpaceName returned "HostSpace"
   name = HostSpace == "HostSpace" = HostSpace : passed
   canAccessFromHost = 1 == true : passed
   pointerAccessibleFromExecutionSpace(ptr, Cuda()): false
   ! canAccessFromCuda = 1 == true : passed
   rawInt == 666 = 1 == true : passed
   Result: true
  Test host Kokkos::View
   memorySpaceName returned "HostSpace"
   name = HostSpace == "HostSpace" = HostSpace : passed
   canAccessFromHost = 1 == true : passed
   pointerAccessibleFromExecutionSpace(ptr, Cuda()): false
   ! canAccessFromCuda = 1 == true : passed
   intView() == 418 = 1 == true : passed
   Result: true
  Test raw cudaMalloc
   ! canAccessFromHost = 1 == true : passed
   canAccessFromCuda = 1 == true : passed
   memorySpaceName returned "CudaSpace"
   name = CudaSpace == "CudaSpace" = CudaSpace : passed
   Result: true
  Test Kokkos::View of CudaSpace
 
 p=0: *** Caught standard std::exception of type 'std::runtime_error' :
 
  cudaGetLastError() error( cudaErrorInvalidValue): invalid argument /home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_CudaExec.hpp:401
  Traceback functionality not available
  
 [FAILED]  (0.00349 sec) Utils_CheckMemoryType_UnitTest
 Location: /home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/tpetra/core/test/Utils/CheckMemoryType.cpp:224
```
Not sure what's up here.

**Comment by mhoemmen** [comment]
Hm, Intel build is funny.  Let me try again.

**Comment by mhoemmen** [comment]
Let's try this again.  My CUDA 9.2 build passes Tpetra's tests just fine.  Not sure what's going on with the PR tester's CUDA 9.2 build, but we'll try it again.

**Comment by mhoemmen** [comment]
@bathmatt Since I had to fix something anyway, I took the opportunity in the commits I just added, to add input checking to `Tpetra::Map` and `Tpetra::MultiVector`.  I will need to add checking to other Tpetra classes, like CrsGraph (which seems to be involved in the Panzer issue somehow), but this is a good start.

**Comment by mhoemmen** [comment]
@trilinos/framework Did the RETEST flag trigger two builds?  Weird.

**Comment by jwillenbring** [comment]
> Something appears to be wrong with tests.

It seems the issue was a temporary issue with CDash.

**Comment by prwolfe** [comment]
Good to know – I was not seeing anything this morning and was wondering.

-----------------------------------------------------------------------
Recommending policy and creating automation to
enable DevOps in the Trilinos and CTH organizations.

**Comment by jwillenbring** [comment]
@mhoemmen 

> Did the RETEST flag trigger two builds? Weird.

This is only a guess, but I have seen this kind of behavior before when a test run has had issues reporting back to GitHub (so there is no failure comment logged), however the autotester still knows  of the failure and will retest as usual. This seems particularly likely since there are CDash results up for both builds.

**Comment by mhoemmen** [comment]
Summary of test failures:

### CUDA build
```
 Starting CheckMemoryType test
  Test raw integer on host
   memorySpaceName returned "HostSpace"
   name = HostSpace == "HostSpace" = HostSpace : passed
   canAccessFromHost = 1 == true : passed
   pointerAccessibleFromExecutionSpace(ptr, Cuda()): false
   ! canAccessFromCuda = 1 == true : passed
   rawInt == 666 = 1 == true : passed
   Result: true
  Test host Kokkos::View
   memorySpaceName returned "HostSpace"
   name = HostSpace == "HostSpace" = HostSpace : passed
   canAccessFromHost = 1 == true : passed
   pointerAccessibleFromExecutionSpace(ptr, Cuda()): false
   ! canAccessFromCuda = 1 == true : passed
   intView() == 418 = 1 == true : passed
   Result: true
  Test raw cudaMalloc
   ! canAccessFromHost = 1 == true : passed
   canAccessFromCuda = 1 == true : passed
   memorySpaceName returned "CudaSpace"
   name = CudaSpace == "CudaSpace" = CudaSpace : passed
   Result: true
  Test Kokkos::View of CudaSpace
 
 p=0: *** Caught standard std::exception of type 'std::runtime_error' :
 
  cudaGetLastError() error( cudaErrorInvalidValue): invalid argument /home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2@2/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_CudaExec.hpp:401
  Traceback functionality not available
  
 [FAILED]  (0.00256 sec) Utils_CheckMemoryType_UnitTest
 Location: /home/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2@2/Trilinos/packages/tpetra/core/test/Utils/CheckMemoryType.cpp:224
 ```
I don't get this failure in my CUDA 9.2 build.  Let me try to replicate this first.  @bathmatt & co., if you are in the mood, please feel free to help.  I've added commits that will check user errors in `Tpetra::Map` and `Tpetra::MultiVector` constructor input.  The latter has shown up before in Drekar, Panzer, or whatever it's called, so I'm hoping it will help again.

### GCC 4.8.4 with ancient OpenMPI

Likely `Anasazi_Epetra_OrthoManagerGenTester_0_MPI_4` failure.  (I didn't actually change the implementation of `Teuchos::ScalarTraits<std::complex<T>>`.)
```
projectAndNormalizeGen() returned rank 3
   || <S,S> - I || after  : 1.05981e-12
         vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv         tolerance exceeded! test failed!
  19|| S_in - X1*C1 - X2*C2 - S_out*B || : 1.06971e-08
  19|| <Y[0],S> || after      : 2.91827e-12
```

**Comment by mhoemmen** [comment]
I think I know what's going on here.  In CUDA 9.2, the only way `cudaPointerGetAttributes` can identify an unregistered host pointer is by returning `cudaErrorInvalidValue`.  That's an error code other than `cudaSuccess`.  This must stick around as the "last error" that CUDA encountered.  My code didn't clear the error by calling `cudaGetLastError`, so the next time I did some Kokkos thing, Kokkos might have thought that CUDA was returning an error when it really wasn't.

This suggests I can fix this issue by calling `cudaGetLastError` after calling `cudaPointerGetAttributes`.  I will do that.

**Comment by mhoemmen** [comment]
I was able to replicate the test failure in a debug CUDA build.  When I added the `cudaGetLastError` call, the test passed.

**Comment by mhoemmen** [comment]
Let's try this again.

**Comment by mhoemmen** [comment]
IT PASSED YAY!!!

**Comment by mhoemmen** [comment]
Thanks @tjfulle ! :-D
